### PR TITLE
feat: Create Pydantic models for Search Pipeline Run API

### PR DESCRIPTION
**Pull request by yuechao-qin**
### TL;DR

Added a new `filter_query` parameter to the pipeline runs API with Pydantic validation models, while maintaining backward compatibility with the existing `filter` parameter.

### What's new?

JSON definition for `filter_query`:

```json
{
  // logical (any predicate)
  "<and|or>": [
    // leaf
    {"key_exists": {"key": "<KEY>"}},
    // leaf
    {"value_contains": {"key": "<KEY>", "value_substring": "<VALUE>"}},
    // leaf
    {"value_in": {"key": "<KEY>", "values": ["<V1>", "<V2>", "..."]}},
    // leaf
    {"value_equals": {"key": "<KEY>", "value": "<VALUE>"}},
    // leaf
    {"time_range": {"key": "system/pipeline_run.date.created_at", "start_time": "<START_DATE>", "end_time": "<END_DATE>"}},
    // logical (leaf only)
    {"not": {"<LEAF_PREDICATE>": {}}},
    // logical (any predicate)
    {"<and|or>": [...]},
    ...
  ]
}
``` 

Example JSON for `filter_query`:

```json
{
  "and": [
    {"key_exists": {"key": "team"}},
    {"value_equals": {"key": "env", "value": "prod"}},
    {"value_in": {"key": "region", "values": ["us-east", "us-west", "eu-west"]}},
    {"not": {"key_exists": {"key": "deprecated"}}},
    {"or": [{"value_contains": {"key": "name", "value_substring": "nightly"}}]}
  ]
}
```

### What changed?

API `GET /api/pipeline_runs/`

#### Functional

- New `filter_query` if used in API will return an HTTP 501 (unimplemented error)
- Only `filter` or `filter_query` (mutual exclusive) can be in the API, else it returns a HTTP 422.

#### Other

- Added a new `filter_query` parameter to `ListPipelineRunsParams` that accepts structured JSON queries
- Created comprehensive Pydantic models in `filter_query_models.py` for validating filter queries, including:
  - Leaf predicates: `key_exists`, `value_equals`, `value_contains`, `value_in`, `time_range`
  - Logical operators: `and`, `or`, `not`
  - Recursive nesting support for complex queries
- Implemented JSON validation for `filter_query` with proper error handling
- Enforced strict validation rules including non-empty strings and timezone-aware datetimes

### How to test?

```bash
uv run pytest tests/test_api_server_sql.py tests/test_filter_query_models.py
```

- Test the new `filter_query` parameter with valid JSON structures like `{"and": [{"key_exists": {"key": "team"}}]}`
- Verify that using both `filter` and `filter_query` simultaneously returns a 422 error
- Confirm that invalid JSON in `filter_query` triggers proper validation errors
- Check that valid `filter_query` usage returns a 501 "not yet implemented" response
- Test edge cases like empty strings, missing timezone info, and nested logical operators

### Why make this change?

This PR introduces the Pydantic models and API wiring that upstream PRs depend on to implement the Search Pipeline Run API.

### CODE REVIEW

**Comment by yuechao-qin** [comment]
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.com/github/pr/TangleML/tangle/106?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#106** <a href="https://app.graphite.com/github/pr/TangleML/tangle/106?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈 <a href="https://app.graphite.com/github/pr/TangleML/tangle/106?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* **#105** <a href="https://app.graphite.com/github/pr/TangleML/tangle/105?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `master`


<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.

File: cloud_pipelines_backend/filter_query_models.py [hunk 1]
```diff
@@ -0,0 +1,118 @@
+from __future__ import annotations
+
+from typing import Annotated
+
+from pydantic import AwareDatetime, BaseModel, Field, StringConstraints, model_validator
+
+NonEmptyStr = Annotated[str, StringConstraints(min_length=1)]
+
+
+class _StrictBase(BaseModel):
```
Comment (yuechao-qin) [inline cloud_pipelines_backend/filter_query_models.py]: Rename this to be `_BaseModel` and use it everywhere!

File: cloud_pipelines_backend/api_server_sql.py [hunk 2]
```diff
@@ -6,7 +6,7 @@
 import typing
 from typing import Annotated, Any, Final, Optional
 
-from fastapi import Query
+from fastapi import HTTPException, Query
```
Comment (yuechao-qin) [inline cloud_pipelines_backend/api_server_sql.py]: Should not have fastApi in this module, use apirouter to define error, to convert custom exception to http exceptions.

nothing http related should be here.

File: cloud_pipelines_backend/api_server_sql.py [hunk 3]
```diff
@@ -33,6 +33,7 @@ def _get_current_time() -> datetime.datetime:
 from . import backend_types_sql as bts
 from . import errors
 from .errors import ItemNotFoundError
+from .filter_query_models import FilterQuery
```
Comment (yuechao-qin) [inline cloud_pipelines_backend/api_server_sql.py]: only import modules, not classes/vars/etc.

File: cloud_pipelines_backend/filter_query_models.py [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+    | AndPredicate
+    | OrPredicate
+)
+
+# Resolve forward reference to "Predicate" in recursive and/or models
+AndPredicate.model_rebuild()
+OrPredicate.model_rebuild()
+
+
+class FilterQuery(_StrictBase):
```
Comment (Volv-G) [inline cloud_pipelines_backend/filter_query_models.py]: nit, but why did we choose to have and/or as the root, and not any predicate?

**Review by Volv-G — APPROVED** [review]
looks good, just 1 question on the design choice

File: cloud_pipelines_backend/filter_query_models.py (hunk 4 above)
Comment (yuechao-qin) [reply cloud_pipelines_backend/filter_query_models.py]: Purely to keep the logic simpler.

Instead of detecting root as:
- Leaf predicate (single item, ~5 variations)
- AND (list)
- OR (list)

Just detecting AND/OR (list) for root was simpler.  Also, once you have more than 1 leaf predicate (5 variations), you need a AND/OR to group them anyways.

File: cloud_pipelines_backend/filter_query_models.py [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+
+
+class ValueIn(_BaseModel):
+    key: NonEmptyStr
+    values: list[NonEmptyStr] = pydantic.Field(min_length=1)
+
+
+class ValueEquals(_BaseModel):
+    key: NonEmptyStr
+    value: NonEmptyStr
```
Comment (Ark-kun) [inline cloud_pipelines_backend/filter_query_models.py]: Hmm. Maybe empty value is OK.

File: cloud_pipelines_backend/filter_query_models.py [hunk 6]
```diff
@@ -0,0 +1,118 @@
...
+
+
+class TimeRange(_BaseModel):
+    """AwareDatetime requires timezone info (e.g. "2024-01-01T00:00:00Z").
+    Naive datetimes like "2024-01-01T00:00:00" are rejected, preventing
+    ambiguous timestamps that could silently resolve to the wrong timezone."""
+
+    key: NonEmptyStr
+    start_time: pydantic.AwareDatetime
```
Comment (Ark-kun) [inline cloud_pipelines_backend/filter_query_models.py]: start_time can be optional too

File: tests/test_filter_query_models.py [hunk 7]
```diff
@@ -0,0 +1,227 @@
+import pydantic
+import pytest
+
+from cloud_pipelines_backend import filter_query_models
+
+
+class TestFilterQuery:
+    def test_full_example_from_design_doc(self):
+        json_str = """
```
Comment (Ark-kun) [inline tests/test_filter_query_models.py:9]: if you want, you can construct Python dict and then use json.dumps to serialize it.
But JSON string like you have is OK too.

File: cloud_pipelines_backend/errors.py [hunk 8]
```diff
@@ -8,3 +8,13 @@ class ItemAlreadyExistsError(Exception):
...
     pass
+
+
+class ApiValidationError(Exception):
+    """Base for all filter/annotation validation errors -> 422."""
+
+    pass
+
+
+class MutuallyExclusiveFilterError(ApiValidationError):
```
Comment (Ark-kun) [inline cloud_pipelines_backend/errors.py]: `MutuallyExclusiveFilterError` is pretty specific to pipeline run search. Maybe it should be declared in that module? Or you can just use `ApiValidationError`
Or if you envision more APIs with mutually exclusive parameters (we avoid this), you could rename the class to `MutuallyExclusiveParametersError`.

Comment (yuechao-qin) [reply cloud_pipelines_backend/errors.py]: Wasn't sure explicit Errors were preferred.

Will just use `ApiValidationError`.

File: tests/test_filter_query_models.py (hunk 7 above)
Comment (yuechao-qin) [reply tests/test_filter_query_models.py:9]: I wanted to use a string to be clear for the developer/reviewer what the JSON looks like as a whole, since this is for testing.
